torch/csrc/jit/codegen/cuda/scheduler_registry.cpp

```diff
@@ -0,0 +1,320 @@
+#include <torch/csrc/jit/codegen/cuda/scheduler_registry.h>
+#include <torch/csrc/jit/codegen/cuda/ir_utils.h>
+#include <torch/csrc/jit/codegen/cuda/root_domain_map.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+bool SchedulerEntry::sameAs(const SchedulerEntry* other) {
+  if (has_param_ != other->has_param_) {
+    return false;
+  }
+  if (has_param_) {
+    return rparams_ == other->rparams_;
+  }
+  return true;
+}
+
+namespace {
+inline bool isTrivialReduction(ReductionOp* red) {
+  auto o_tv = red->out()->as<TensorView>();
+  // Assuming graph unscheduled at this point.
+  for (auto id : o_tv->getRootDomain()) {
+    if (id->isReduction() && !id->rawExtent()->isOneInt()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+std::vector<ReductionOp*> findReductionOps(Fusion* fusion) {
+  std::vector<ReductionOp*> red_ops;
+  for (auto expr : fusion->exprs()) {
+    if (auto red = dynamic_cast<ReductionOp*>(expr)) {
+      if (!isTrivialReduction(red)) {
+        red_ops.push_back(red);
+      }
+    }
+  }
+  return red_ops;
+}
+
+std::vector<TensorView*> findOutputsOfRed(Fusion* fusion, TensorView* red_tv) {
+  TORCH_INTERNAL_ASSERT(fusion->inFusion(red_tv));
+  auto output_set = DependencyCheck::getAllOutputsOf({red_tv});
+  auto tv_entries = ir_utils::filterByType<TensorView>(output_set);
+  std::vector<TensorView*> tv_outputs_of_reduction(
+      tv_entries.begin(), tv_entries.end());
+  return tv_outputs_of_reduction;
+}
+
+} // namespace
+
+namespace {
+
+class SingleReductionScheduler : public SchedulerEntry {
+ public:
+  explicit SingleReductionScheduler(Fusion* fusion, ExpressionEvaluator& ee)
+      : SchedulerEntry(ScheduleHeuristic::Reduction, true) {
+    getHeuristics(fusion, ee);
+  }
+
+  //! Check if the reduction heuristics apply in given fusion
+  static bool canSchedule(Fusion* fusion) {
+    auto red_ops = findReductionOps(fusion);
+    if (red_ops.size() != 1) {
+      return false;
+    }
+
+    auto red_tv = red_ops[0]->out()->as<TensorView>();
+
+    // Not allowing broadcasting reduction result to support
+    //  grid reduction. This is an overkill might want to consider
+    //  trying to get the heuristics and check only if grid reduction is
+    //  required.
+    auto uses = DependencyCheck::getAllUseChains(red_tv);
+    for (auto& chain : uses) {
+      for (auto val : chain) {
+        if (val->definition()->isA<BroadcastOp>()) {
+          return false;
+        }
+      }
+    }
+
+    return true;
+  }
+
+  void schedule(Fusion* fusion) override {
+    // TODO find outputs of tv: what would we need to fill in?
```

Comment:
please add a perf scope here (same for the rest of the ::schedule() implementations...)

Reply:
Updated, thanks!